packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_InsertGlobalIndices_FE_SP.hpp

```diff
@@ -272,6 +277,226 @@ int executeInsertGlobalIndicesFESP_(const Teuchos::RCP<const Teuchos::Comm<int>
+  // - sumIntoGlobalValues( 3,  [  2  3  7  6  ],  [  -1  2  -1  0  ])
+  // - sumIntoGlobalValues( 7,  [  2  3  7  6  ],  [  0  -1  2  -1  ])
+  // - sumIntoGlobalValues( 6,  [  2  3  7  6  ],  [  -1  0  -1  2  ])
+  RCP<TimeMonitor> timerElementLoopMemory = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("3.1) ElementLoop  (Memory)")));
+
+  RCP<fe_matrix_t> fe_matrix = rcp(new fe_matrix_t(fe_graph));
+  RCP<fe_multivector_t> rhs =
+    rcp (new fe_multivector_t(domain_map, fe_graph->getImporter(), 1));
+
+  auto localMatrix  = fe_matrix->getLocalMatrix();
+  auto localRHS     = rhs->getLocalViewDevice();
+  auto localMap     = owned_plus_shared_map->getLocalMap();
+  auto localColMap  = fe_matrix->getColMap()->getLocalMap();
+
+  // Because we're processing elements in parallel, we need storage for all of them
+  int numOwnedElements = mesh.getNumOwnedElements();
+  int nperel = owned_element_to_node_ids.extent(1); 
+  pair_type alln = pair_type(0,nperel);
+  scalar_2d_array_t all_element_matrix("all_element_matrix",nperel*numOwnedElements);
+  scalar_1d_array_t all_element_rhs("all_element_rhs",nperel*numOwnedElements);
+  local_ordinal_view_t  all_lcids("all_lids",nperel*numOwnedElements);
+
+  timerElementLoopMemory=Teuchos::null;
+  RCP<TimeMonitor> timerElementLoopMatrix = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("3.2) ElementLoop  (Matrix)")));
+
+  // Loop over elements
+  Tpetra::beginFill(*fe_matrix,*rhs);
+  Kokkos::parallel_for(Kokkos::RangePolicy<execution_space_t>(0, numOwnedElements),KOKKOS_LAMBDA(const size_t& element_gidx) {
+      // Get subviews      
+      pair_type location_pair = pair_type(nperel*element_gidx,nperel*(element_gidx+1));
+      auto element_rhs    = Kokkos::subview(all_element_rhs,location_pair);
+      auto element_matrix = Kokkos::subview(all_element_matrix,location_pair,alln);
+      auto element_lcids  = Kokkos::subview(all_lcids,location_pair);
+
+      // Get the contributions for the current element
+      ReferenceQuad4(element_matrix);
+      ReferenceQuad4RHS_Kokkos(element_rhs);
+
+      // Get the local column ids array for this element
+      for(int element_node_idx=0; element_node_idx<nperel; element_node_idx++) {
+        element_lcids(element_node_idx) = localColMap.getLocalElement(owned_element_to_node_ids(element_gidx, element_node_idx));
+      }
+      
+      // For each node (row) on the current element:
+      // - populate the values array
+      // - add the values to the fe_matrix.
+      for(int element_node_idx=0; element_node_idx<nperel; element_node_idx++)
+        {
+          local_ordinal_t local_row_id = localMap.getLocalElement(owned_element_to_node_ids(element_gidx, element_node_idx));
+          local_ordinal_t local_col_id = localColMap.getLocalElement(owned_element_to_node_ids(element_gidx, element_node_idx));
+          auto row_values = Kokkos::subview(element_matrix,element_node_idx,alln);
+          // Force atomics on sums
```

Comment:
Consider not using atomics for `Kokkos::Serial`.  This will make comparisons with OpenMP and CUDA more fair.

Reply:
Kokkos::atomic_add is supposed to be non-atomic for Kokkos::Serial.

Reply:
@csiefer2 wrote:
> `Kokkos::atomic_add` is supposed to be non-atomic for `Kokkos::Serial`.

I think that's only true if OpenMP and CUDA are disabled as build options.  `Kokkos::atomic_*` functions don't know the current execution space.